### DISCUSSION TITLE
Status of Opencrx BPI Adapter

### DISCUSSION THREAD

**Discussion by nyemiks** [opencrx/opencrx#145] [Q&A]
I recently discovered that opencrx has a Business Process Integration (BPI) adapter.
I will like to know more about it ? What are the use cases for using it ? is it still supported ? How is it different from using the swagger opencrx rest api ? Can It be extended ?

is there some existing documentation about how to use it ?  i.e sample request and response format for testing the bpi adpter.

There is a possibility that I will need to integrate opencrx with a  Business Process Management System (BPMS) for a client. So I am curious to know how this BPI adapter works.

**Comment by wfro64** [comment]
Yes, it is still supported. Unfortunately, there is not much documentation. All relevant classes are in package https://github.com/opencrx/opencrx/tree/master/core/src/main/java/org/opencrx/application/bpi.

The BpiAdapter allows a simple, custom-specific extension of the REST-API of openCRX based on the existing UML model and backend logic. 

Custom-actions are added by extending the class BpiAdapterServlet:
- Override the method BpiAdapterServlet.init() and register actions with registerAction()
- Extend the class BpiPlugin with additional custom-logic and override the method BpiAdapterServlet.newPlugIn() returning the extended plug-in

An action has full access to the openCRX-API. You can find samples in the package https://github.com/opencrx/opencrx/tree/master/core/src/main/java/org/opencrx/application/bpi/adapter.
